### PR TITLE
feat: Add homebrew metadata

### PR DESCRIPTION
**Pull request by Avgor46**
Implements #1619

### CODE REVIEW

File: annotator/misc/brewsource/brewsource.go [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+		if pkg.PURLType != purl.TypeBrew {
+			continue
+		}
+		md, ok := pkg.Metadata.(*metadata.Metadata)
+		if !ok {
+			continue
+		}
+		var repoURL string
+		// Try md.URL, then md.Head, then each md.Mirrors until one yields a valid repo URL.
+		candidates := make([]string, 0, 2+len(md.Mirrors))
```
Comment (another-rex) [inline annotator/misc/brewsource/brewsource.go]: would this work?

```suggestion
		candidates := append([]string{md.URL, md.Head}, md.Mirrors...)
```

File: annotator/misc/brewsource/brewsource.go [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+	Name = "misc/brew-source"
+)
+
+var vcsPrefixes = [2]string{
+	"https://github.com",
+	"https://gitlab.com",
+}
+
+// Annotator adds repository source code identifiers for extracted Homebrew packages.
+type Annotator struct{}
```
Comment (another-rex) [inline annotator/misc/brewsource/brewsource.go:43]: Please mention here a quick summary of how this does the annotation. E.g. Does a best effort match from the download url using x y z fields,

File: annotator/misc/brewsource/brewsource.go [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+			continue
+		}
+		// extract the repository URL, for example https://github.com/owner/repo.git
+		urlParts := strings.SplitN(url, "/", 6)
+		if len(urlParts) < 5 {
+			return "", fmt.Errorf("invalid VCS URL: %s", url)
+		}
+		repoURL := fmt.Sprintf("%s/%s/%s", prefix, urlParts[3], urlParts[4])
+		if !strings.HasSuffix(repoURL, ".git") {
+			repoURL += ".git"
```
Comment (another-rex) [inline annotator/misc/brewsource/brewsource.go]: No need for the .git suffix.

File: annotator/misc/brewsource/brewsource.go [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+
+// Name returns the name of the annotator.
+func (Annotator) Name() string { return Name }
+
+// Version returns the version of the annotator.
+func (Annotator) Version() int { return 0 }
+
+// Requirements returns the requirements of the annotator.
+func (Annotator) Requirements() *plugin.Capabilities {
+	return &plugin.Capabilities{OS: plugin.OSMac, RunningSystem: true}
```
Comment (another-rex) [inline annotator/misc/brewsource/brewsource.go]: I don't think this plugin in particular needs a RunningSystem, nor technically needs a specific OS either, so can remove both.

The actual requirements are on the initial extraction.

File: extractor/filesystem/os/homebrew/homebrew.go [hunk 5]
```diff
@@ -99,30 +104,81 @@ func (e Extractor) Extract(ctx context.Context, input *filesystem.ScanInput) (in
...
+		pkg.Metadata = parseFormula(filepath.Join(input.Root, p.AppFormulaPath))
+	}
+
+	return inventory.Inventory{Packages: []*extractor.Package{pkg}}, nil
 }
 
 // SplitPath takes the package path and splits it into its recognised struct components
-func SplitPath(path string) *BrewPath {
-	path = strings.ToLower(path)
+func SplitPath(origPath string) *BrewPath {
```
Comment (another-rex) [inline extractor/filesystem/os/homebrew/homebrew.go]: Since this is now a bit more than just a path, let's rename BrewPath to BrewPackage.

File: extractor/filesystem/os/homebrew/homebrew.go [hunk 6]
```diff
@@ -99,30 +104,81 @@ func (e Extractor) Extract(ctx context.Context, input *filesystem.ScanInput) (in
...
+			return &brewPackage
 		}
 	}
 	return nil
 }
+
+// Function parses Cellar/*/*/.brew/*.rb files to extract URL, Head, and Mirrors if possible.
+func parseFormula(formulaFilePath string) *metadata.Metadata {
+	metadata := &metadata.Metadata{}
+	file, err := os.Open(formulaFilePath)
```
Comment (another-rex) [inline extractor/filesystem/os/homebrew/homebrew.go]: Please use the virtual filesystem provided by `input ScanInput` to open the file.

Comment (Avgor46) [reply extractor/filesystem/os/homebrew/homebrew.go]: done

File: extractor/filesystem/os/homebrew/homebrew.go (hunk 5 above)
Comment (Avgor46) [reply extractor/filesystem/os/homebrew/homebrew.go]: renamed

File: annotator/misc/brewsource/brewsource.go (hunk 4 above)
Comment (Avgor46) [reply annotator/misc/brewsource/brewsource.go]: done

File: annotator/misc/brewsource/brewsource.go (hunk 3 above)
Comment (Avgor46) [reply annotator/misc/brewsource/brewsource.go]: done

File: annotator/misc/brewsource/brewsource.go (hunk 2 above)
Comment (Avgor46) [reply annotator/misc/brewsource/brewsource.go:43]: done

File: annotator/misc/brewsource/brewsource.go (hunk 1 above)
Comment (Avgor46) [reply annotator/misc/brewsource/brewsource.go]: done

**Comment by another-rex** [comment]
Seems like there's two minor issues.

1. We have a check to block `.` and other edge case characters in filenames so avoid issues on specific OSes. Since it's important for the file path to be correct as part of the test, can the file path be generated at runtime?
2. Windows test seems to be failing, I think we can skip the test if it's on windows (if , since the extractor is only supported on unix systems anyway.

e.g. https://github.com/google/osv-scalibr/blob/153bd694756cd89c3a45f0f8df0fd4b9f3cf11b7/detector/cis/generic_linux/etcpasswdpermissions/etcpasswdpermissions_test.go#L40-L43

File: extractor/filesystem/os/homebrew/homebrew.go [hunk 7]
```diff
@@ -99,30 +104,81 @@ func (e Extractor) Extract(ctx context.Context, input *filesystem.ScanInput) (in
...
+			brewPackage := BrewPackage{
 				AppName:    pathParts[i+1],
 				AppVersion: pathParts[i+2],
 				AppFile:    pathParts[i+3],
 				AppExt:     pathParts[len(pathParts)-1],
 			}
+			// For Cellar packages, attempt to extract the formula file path.
+			if pathPart == cellarPath {
+				origPathParts := strings.Split(origPath, "/")
+				formulaFilePath := filepath.Join(strings.Join(origPathParts[:i+3], "/"), ".brew", brewPackage.AppName+".rb")
```
Comment (another-rex) [inline extractor/filesystem/os/homebrew/homebrew.go]: Use path instead of filepath when doing the join, as we are using the virtual FS which always uses `/`.

Comment (Avgor46) [reply extractor/filesystem/os/homebrew/homebrew.go]: done

**Comment by Avgor46** [comment]
1. I don't see any CI checks blocking `.` in filenames. The pipeline has now passed in my fork. The previous CI lint problem was with the EOF of rclone.rb file.
2. Homebrew tests on Windows are now disabled.

**Review by another-rex — APPROVED** [review]
